### PR TITLE
Feature/darksteel pickaxe

### PR DESCRIPTION
**Pull request by CrazyPants**
# Description

Base implementation of Dark Steel tools and upgrades

# Checklist:

- [x] My code follows the style guidelines of this project (.editorconfig, most IDEs will use this for you)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

File: enderio-base/src/main/java/com/enderio/base/common/item/darksteel/upgrades/DarkSteelUpgradeRecipe.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+    @Override
+    public ItemStack assemble(Container pInv) {
+        ItemStack resultItem = pInv.getItem(0).copy();
+        Optional<IDarkSteelUpgradable> target = getUpgradableFromItem(resultItem);
+        Optional<IDarkSteelUpgrade> upgrade = getUpgradeFromItem(pInv.getItem(1));
+        return target.map(upgradable -> upgrade.map(up -> DarkSteelUpgradeable.addUpgrade(resultItem, up)).orElse(ItemStack.EMPTY)).orElse(ItemStack.EMPTY);
+    }
+
+    private Optional<IDarkSteelUpgradable> getUpgradableFromItem(@Nullable ItemStack item) {
```
Comment (justliliandev) [inline enderio-base/src/main/java/com/enderio/base/common/item/darksteel/upgrades/DarkSteelUpgradeRecipe.java]: itemstacks are in 99.9% of all cases never null. The only caller of this method is not on of these rare cases. Pls add a check for item.isEmpty but not for null

File: enderio-base/src/main/java/com/enderio/base/common/item/darksteel/upgrades/DarkSteelUpgradeRegistry.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    public ItemStack writeUpgradeToItemStack(ItemStack stack, IDarkSteelUpgrade upgrade) {
+        CompoundTag rootTag = new CompoundTag();
+        rootTag.putString("name", upgrade.getSerializedName());
+        rootTag.put("data", upgrade.serializeNBT());
+        stack.getOrCreateTag().put(UPGRADE_IN_STACK_KEY, rootTag);
+        return stack;
+    }
+
+    public boolean hasUpgrade(@Nullable ItemStack stack) {
+        if(stack == null || !stack.hasTag()) {
```
Comment (justliliandev) [inline enderio-base/src/main/java/com/enderio/base/common/item/darksteel/upgrades/DarkSteelUpgradeRegistry.java]: remove @nullable and check for isEmpty

File: enderio-base/src/main/java/com/enderio/base/common/item/darksteel/upgrades/DarkSteelUpgradeRegistry.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+            return false;
+        }
+        return stack.getOrCreateTag().contains(UPGRADE_IN_STACK_KEY);
+    }
+
+    public Optional<IDarkSteelUpgrade> readUpgradeFromStack(@Nullable ItemStack stack) {
+        if(stack == null || !stack.hasTag()) {
```
Comment (justliliandev) [inline enderio-base/src/main/java/com/enderio/base/common/item/darksteel/upgrades/DarkSteelUpgradeRegistry.java]: see above

**Review by justliliandev — COMMENTED** [review]
I'm not done, and haven't looked into how the registry works, but I'm going to sleep

File: enderio-base/src/main/java/com/enderio/base/common/item/darksteel/upgrades/DarkSteelUpgradeRecipe.java (hunk 1 above)
Comment (CrazyPants) [reply enderio-base/src/main/java/com/enderio/base/common/item/darksteel/upgrades/DarkSteelUpgradeRecipe.java:65]: The caller of this method actually is one of the 0.1% of those case. The item stack can come from this: pInv.getItem(0), this does return null if there is no item in that inv. slot. This is the case regulary in the matches method.

**Pull request closed without merging**